theano/sandbox/gpuarray/multinomial.py

```diff
@@ -0,0 +1,256 @@
+# TODO test dtype != float32
+import os
+
+import pygpu
+
+import theano
+import theano.sandbox.multinomial
+from theano import Apply, config
+from theano.gof import Op, local_optimizer
+from theano.tensor import NotScalarConstantError, get_scalar_constant_value
+from theano.sandbox import gpuarray
+from .basic_ops import as_gpuarray_variable, infer_context_name
+from .fp16_help import write_w
+from .opt import register_opt, op_lifter
+from .type import gpu_context_type, GpuArrayType
+
+
+class GPUAMultinomialFromUniform(gpuarray.basic_ops.GpuKernelBase, Op):
+    __props__ = ("odtype",)
+    params_type = gpu_context_type
+
+    def __init__(self, odtype):
+        Op.__init__(self)
+        self.odtype = odtype
+
+    def get_params(self, node):
+        return node.outputs[0].type.context
+
+    def c_headers(self):
+        return ['<numpy_compat.h>', 'gpuarray_helper.h']
+
+    def c_header_dirs(self):
+        return [os.path.dirname(__file__), pygpu.get_include()]
+
+    def make_node(self, pvals, unis):
+        assert pvals.dtype == 'float32'
+        assert unis.dtype == 'float32'
+        ctx_name = infer_context_name(pvals, unis)
+
+        pvals = as_gpuarray_variable(pvals, ctx_name)
+        unis = as_gpuarray_variable(unis, ctx_name)
+
+        if pvals.ndim != 2:
+            raise NotImplementedError('pvals ndim should be 2', pvals.ndim)
+        if unis.ndim != 1:
+            raise NotImplementedError('unis ndim should be 1', unis.ndim)
+        if self.odtype == 'auto':
+            odtype = pvals.dtype
+        else:
+            odtype = self.odtype
+        assert odtype == 'float32', odtype
+        if odtype != pvals.dtype:
+            raise NotImplementedError(
+                'GpuMultinomialFromUniform works only if '
+                'self.odtype == pvals.dtype', odtype, pvals.dtype)
```

Comment:
I'm just wondering why do you allow setting odtype, if in the end anything that is not the same as the dtype of pvals will give you an error?